File: jan21/beginner/pow_optimized.py
import sys

# wrapper function to kick off work function
def pow_wrap(base, power):
    # handling incorrect cases
    assert power >= 0
    return pow_work(base, power, current)    

def pow_work(base, power, current):
    if power == 0:
        return current
    if power % 2 == 0:
        return pow_work(base * base, power / 2, current)
    else:
        return pow_work(base, power - 1, current * base)
    
def main():
    print('Please enter in the base number: ')
    base = int(sys.stdin.readline())
    
    print('Please enter in the power: ')
    power = int(sys.stdin.readline())
    
    print('The result of ', base, '^', power, ' is ', pow(base, power))

if __name__ == '__main__':
    main()

File: jan21/beginner/pow_optimized.cpp
/*
 * optimizing pow for tail recursion, time complexity,
 * and possible overflows
 */
#include <iostream>

using namespace std;

/*
 * pow is now O(log2(n))
 */

long pow_work(long base, long power, long current)
{
  if (power <= 0) {
    return current;
  }
  
  if (power % 2 == 0) {
    return pow_work(base * base, power/2, current);
  } else {
    return pow_work(base, power - 1, current * base);

  }
}

// wrapper for our worker function
long pow(long base, long power)
{
  return pow_work(base, power, 1);
}

int main()
{
  long base, power;
  cout << "Please enter in the base number: ";
  cin >> base;
  cout << "Please enter in the power: ";
  cin >> power;

  cout << "The result of " << base << '^' << power << " is " << pow(base, power) << endl;
  return 0;
}

